subprojects/core/src/main/java/org/gradle/api/internal/DefaultDomainObjectCollection.java

```diff
@@ -257,6 +258,19 @@ public void addLater(Provider<? extends T> provider) {
         store.addPending(providerInternal);
     }
 
+    @Override
+    public void addAllLater(Provider<? extends Iterable<T>> provider) {
+        assertMutable();
+        CollectionProviderInternal<T, ? extends Iterable<T>> providerInternal = Cast.uncheckedCast(provider);
+        if (eventRegister.isSubscribed(providerInternal.getElementType())) {
```

Comment:
Hmm, doesn't this mean that if I do something like...

```
// pseudo
ListProperty<Task> listProvider = {
    copyTask,
    syncTask,
    defaultTask
}
container.addAllLater(listProvider)
```

If I have a `withType(Task)` action, the provider is realized and each task is realized. This seems mostly right.
If I have a `withType(Copy)` action, the provider is _not_ realized.  What would ever trigger it?

I think the assumption here is that there can only be a single type in the container or `addAllLater` can only handle a single type in the collection. This is sort of OK for a `NamedDomainObjectCollection` because we know there's a single public type, but I think you could (if you wanted to) have `withType()` configuration actions for particular implementation types:

```
class PublicType {
    String name = getClass().getSimpleName()
}

class SubType extends PublicType {}
class AnotherType extends PublicType {}

def container = project.container(PublicType)

container.withType(PublicType) {
    println "public type " + it
}

container.withType(SubType) {
    println "sub type " + it
}

container.withType(AnotherType) {
    println "another type " + it
}

container.add(new PublicType())
container.add(new SubType())
container.add(new AnotherType())
```

I think this was a little different in the software model where you could only reference things by their public type.

Reply:
It's a good question.  Another way to state it: should `withType()` apply to the type of individual values or of the `Provider`?  We of course have the same problem with `Provider<T>` - the type on the provider may or may not be the actual type of the provided value, but we check against the type of the provider.  Given that, I assumed that a `Provider<Iterable>` should return an upper bound and only be realized if that's the type requested, but that may have already been wrong.

I think this actually still works, because we still fire the rules when we do realize each value, but we won't catch it and realize the provider when `addLater()` or `addAllLater()` are called, which might be a problem.  I'm not sure I understand exactly which use case this eager realization addresses, though.